### ISSUE TITLE
Weird query performance - optimize query execution

### ISSUE THREAD

**Issue by tpcgold** [neo4j/neo4j#13288]
Most resent Neo4j AuraDB

I found that Neo4j executes some simple queries in a very strange way. E.g. the most simple case:
```
MATCH p=(a)-[*]->(b:LABEL3) WHERE b.property = 'test'
    and (a:LABEL1 OR a:LABEL2)
RETURN p
```

will filter all labels first and make a Cartesian product -> this is far from optimal. (and will almost never finish on huge graphs)

Instead of 
```
MATCH p=(a)-[*]->(b:LABEL3) WHERE b.property = 'test'
RETURN p
```

This will filter on the target node and return all paths. This is executed in an instant.

Neo4j should consider moving the "Filter" clause to the end of the execution plan in this case.
The current workaround is using the second query and filtering the labels in my own code (after I got the result), as the first query will execute forever.

**Comment by tpcgold** [comment]
In addition:
`MATCH p=(a)-[*]->(b:LABEL3) WHERE b.property = 'test'
    and (a:LABEL1 OR a:LABEL2)
RETURN p LIMIT 25`
With a limit clause, it does exactly what is proposed above. 

So there is an issue as soon as a certain threshold is reached and it starts using a way less efficient cartesian product.

**Comment by AzuObs** [comment]
Hello @tpcgold, 

Thank you for reaching out to us and reporting this issue. 

The query planner will generate many different ways of solving the same query, and will perform a cost-based comparison between each of these ways. In order to do this, it looks at the operators in each query (cartesian product, filter, etc) as well as statistics about the current state of the database which will contain information such as how many nodes of label LABEL1 exist on disk.

It could well be that based on the statistics of the current of the database, the plan with the cartesian product is the most optimal. Such a plan might indeed scale poorly if the state of the database was to significantly change, but the database is able to detect such scenarios and would force a the query to be recompiled (the new plan would probably not contain a cartesian product).

Generally speaking, cartesian product may be cheap to execute if one or both sides of the cartesian product have very few records which might be the case if the right and left hand side operators of the cartesian product are very selective (eg NodeByLabelScan).

If you would like to pursue this issue further, then would you mind please prefixing your queries with `debug=printcostcomparisons` and sending us the output you see in your `neo4j.log` file? I will then ask a team member to have a look and get back to you.

```
CYPHER debug=printcostcomparisons
EXPLAIN
MATCH p=(a)-[*]->(b:LABEL3) 
WHERE b.property = 'test'
      AND a:LABEL1 OR a:LABEL2
RETURN p
```

Hope you have a nice day

**Comment by tpcgold** [comment]
Hi thanks for the fast response. I would do so if I could: the issue is I'm currently on Aura Free (so I'm not able to provide the logs)
![image](https://github.com/neo4j/neo4j/assets/66625570/3133bf02-2496-4e29-83ef-bf5271c8f484)


 In the meantime I surprisingly found that
`MATCH p=(b:LABEL3)<-[*]-(a) WHERE b.property = 'test'
    and (a:LABEL1 OR a:LABEL2)
RETURN p`
does execute while 

`MATCH p=(a)-[*]->(b:LABEL3) WHERE b.property = 'test'
    and (a:LABEL1 OR a:LABEL2)
RETURN p`
will loop forever. the execution plan is identical (so still with Cartesian product). The only difference is that a and b positions are exchanged in the execution graph which should not impact the performance at all.
(I know AxB≠BxA but in the Neo4j case above it's basically a bijective cartesian product)

**Comment by AzuObs** [comment]
Sorry, I forgot that on Neo4j Aura Free the logs are not available. Neo4j can be downloaded and installed locally via this link https://neo4j.com/download/. Installing the software would give you full control over your database. If you would like to build a reproducible case that you think will yield a poor plan, then I'd invite you to give us some seed data and send us the logged cost comparisons.

If I populate this seed data:

```
UNWIND range(0, 100000) AS i
CREATE (:LABEL1)-[:FOO]->(:LABEL3 { property: 'test' })
```

Followed by your original query:

```
EXPLAIN
MATCH p=(a)-[*]->(b:LABEL3) 
WHERE b.property = 'test'
      AND a:LABEL1 OR a:LABEL2
RETURN p
```

Then I get the following plan which does not contain a CartesianProduct.

![plan](https://github.com/neo4j/neo4j/assets/9439499/327b083d-a420-44c6-9248-6fd30a2bad42)

_PS. Different machine might get different plans, because producing a logical plan is not entirely deterministic. But generally speaking for a database with lots of data, plans containing CartesianProduct should be rarer._

**Other issue**

I tried to replicate this issue but was not able to. For me, using the seed data I use above, both queries complete without hanging. Not sure what is happening on your end but if you are able to provide us with a reproducible case then we would be able to investigate further. If this is an avenue you'd like to explore then it might be better to separate this issue into another Github Issue to focus each conversation on a single problem.

```
MATCH p=(b:LABEL3)<-[*]-(a) 
WHERE b.property = 'test' 
  AND (a:LABEL1 OR a:LABEL2) 
RETURN COUNT(*)
// 100001

MATCH p=(a)-[*]->(b:LABEL3)
WHERE b.property = 'test'
  AND (a:LABEL1 OR a:LABEL2)
RETURN COUNT(*)
// 100001
```

I hope these answers were helpful 🙇

**Comment by arnefischereit** [comment]
Hi @tpcgold,

We have not heard from you in the last 3 weeks, so I will assume that Daniel's last comment helped you solve the problem and I will therefore close the issue.

Should you still face the same problem and found a way to replicate it, please feel free to reopen this issue or create a new one.

Regards,

Arne

**Comment by tpcgold** [comment]
Hi, @arnefischereit haven't had time to create a minimum example and at least one way of CQL did work. I've just checked that the problem still occurs. This does not work:
`MATCH p=(a)-[*]->(b:LABEL3) WHERE b.property = 'test' and (a:LABEL1 OR a:LABEL2) RETURN p`

Here is an anonymized form of the (real) execution plan:
![image](https://github.com/neo4j/neo4j/assets/66625570/a65cfeff-6ca2-4c48-add1-f2195a4e87c4)


while this does (within seconds):
`MATCH p=(b:LABEL3)<-[*]-(a) WHERE b.property = 'test' and (a:LABEL1 OR a:LABEL2) RETURN p`
Anonymized execution plan:
![image](https://github.com/neo4j/neo4j/assets/66625570/d987cd06-d98d-4e2a-b595-4f5eda780b63)

You can see it's basically the same but interchanged. So I don't get why this is such a huge difference as it's interchangeable and the Neo4j query-optimizer should find the optimal execution (which isn't the case)

So Neo4j should probably consider this as Query-Optimizer Bug

**Comment by arnefischereit** [comment]
Hi @tpcgold,

I would assume that the two plans perform the same. Could you therefore not only attach the output from `EXPLAIN` but also from `PROFILE`, so that we can see what is taking long? Also, it would be great if you could attach that in text format. You can find an option to export as TXT here:
![image](https://github.com/neo4j/neo4j/assets/79841228/f30937e3-b47e-4d33-9ab4-53ad28436016)

And to be able to reproduce what plans are generated on your side, please also provide the output from `CALL db.stats.retrieve('GRAPH COUNTS')`, which will give us the relevant meta-data about your database.

Cheers,

Arne

**Comment by arnefischereit** [comment]
Hi @tpcgold,

This issue has been idle for another few weeks. Does the comment I made make sense to you? Is this still on your agenda?

If you are still experiencing this issue, it would be great if you can attach the information we need to look into this. Otherwise, we would like to close this issue and you can open a new one once you have the data ready.

**Comment by tpcgold** [comment]
Hi @arnefischereit yes it totally makes sense. I am very sorry it's taking this long but I haven't found time to work on the ticket. (I'll add the `Profile` in the upcoming week)

-> especially first query takes forever (never finished on the full dataset while the second did), so I have to investigate to what extent I can reduce the data so it will finish and create a profile that shows the problem

**Comment by tpcgold** [comment]
@arnefischereit Finally I was able to run the Profile. (I had to go down to "limit 40" so the problematic plan did run)
It clearly shows in the TXT file what's going on in the wrongly executed query. (again I anonymized the labels keeping everything else intact)
[working_plan.txt](https://github.com/neo4j/neo4j/files/13173740/working_plan.txt)
[problem_plan.txt](https://github.com/neo4j/neo4j/files/13173741/problem_plan.txt)

 

Unfortunately, I'm not able to run 
`CALL db.stats.retrieve('GRAPH COUNTS')`
on Neo4j Aura Free (seems this Admin Call doesn't work there due to permissions)
but maybe this is what you want to see (the DB is far from full):
![2023-10-26_08-25](https://github.com/neo4j/neo4j/assets/66625570/bf60fff2-01b3-43a2-91e2-d7626485308c)

**Comment by tpcgold** [comment]
Just that you know what I mean with "it clearly shows the problem":

I mean step ID 4 in both files 
while on the **working plan** the CartesioanProduct is constantly 5000
on the **problematic plan** the row numbers are exploding the higher the limit is set

(e.g. switching the limit from 40 to 41 is enough to see this)
[working_plan_limit41.txt](https://github.com/neo4j/neo4j/files/13174020/working_plan_limit41.txt)
[problem_plan_limit41.txt](https://github.com/neo4j/neo4j/files/13174021/problem_plan_limit41.txt)

(more than 41 is not possible on the problem plan and less than 40 doesn't show the problem as it's a different "linear" execution plan without cartesian product)

**Comment by arnefischereit** [comment]
Hi @tpcgold,

Thank you for coming back to this and sorry that retrieval of the graph counts is not available in Aura Free. I can see two topics in this issue, so I will address them individually:

## Order of the Cartesian product

It looks to me like there are far fewer relationships we need to traverse from `t` for finding the paths. This could for example be because `u` is more connected than `t` such as in this simplified sketch:
![Untitled graph(2)](https://github.com/neo4j/neo4j/assets/79841228/c5b3b897-a08d-423d-86d0-3f1f9ea7cb44)

That should explain why one plan performs a lot better than the other.

## Getting a linear plan

In the original post, you wrote that

```Cypher
MATCH p=(a)-[*]->(b:LABEL3) WHERE b.property = 'test'
RETURN p
```

performs better as it produces a linear plan and does not use a Cartesian product.

Usually, it helps giving the planner more information on the relationships, such as applying relationship types. That way, it may become better at estimating when a Cartesian product is preferable to a linear plan.

In case, you cannot make it work, you can "force" the planner by introducing a `WITH * SKIP 0` to separate the parts of the query which should be solved first and which after that:

```Cypher
MATCH p=(a)-[*]->(b:LABEL3) WHERE b.property = 'test'
WITH * SKIP 0 WHERE a:LABEL1|LABEL2
RETURN p
```

This should lead to a linear plan.

## Conclusion

Without looking at your data in more detail, there is not much more I can help you tweak the query. But I hope this will provide you with the necessary tools you need to get the performance that you want.

Let me know if there is more I can help you with.

**Comment by tpcgold** [comment]
Wow thanks for our response the 
`WITH * SKIP 0 WHERE a:LABEL1|LABEL2`
did the trick. 

this is a huge difference the "problem" query now executes in milliseconds and even the "working" query is now faster.
pretty interesting the query optimizer doesn't plan this way and isn't using this path.

beside this, I would also say even if:
> It looks to me like there are far fewer relationships we need to traverse from t for finding the paths. 

this should be automatically recognized by the query optimizer (neo4j internally).
so it automatically selects the optimal execution plan (e.g. query in one or another direction depending on what is more efficient)

**so although we have a workaround now, I still think this needs to be properly addressed**:
As any Neo4j User wants the same and most optimal query performance (no matter which direction he writes the query - as they are interchangeable and it's the query optimizers job to find the best/fastest path)

**Comment by arnefischereit** [comment]
Hi @tpcgold,

Good to hear that this solved your issue. I do agree that in general, the perfect plan should be chosen.

However, there are cases, where we have too little information on the pattern that the user is looking for to generate an appropriate plan. Keeping track on all possible patterns is equivalent to caching the database: For each pair of `u` and `t` we would need to remember how many paths there are of a certain kind between the two.

To investigate whether we could have been smarter in this particular case, we would the graph counts that I mentioned earlier. I understand that this is not possible for you to produce via the procedure. But if we should investigate, maybe you could send me the details from:

```
COUNT { () }
```

```
MATCH (n)
UNWIND labels(n) AS label
RETURN label, count(*)
```

```
MATCH ()-[r]->()
RETURN type(r), count(*)
```

```
MATCH (a)-[r]->()
UNWIND labels(a) AS label
WITH label, type(r) as relType
RETURN [label, relType], count(*);
```

```
MATCH ()-[r]->(a)
UNWIND labels(a) AS label
WITH label, type(r) as relType
RETURN [relType, label], count(*);
```

That might give us enough clues about what the optimizer's decisions are in your case.

**Comment by tpcgold** [comment]
Hi @arnefischereit here a consistently anonymised version of the outputs the above queries:

`COUNT { () }` <- doesn't exist 
you mean 
`MATCH () RETURN count(*) as count` ?
it's `6857`


```
MATCH (n)
UNWIND labels(n) AS label
RETURN label, count(*)
```
[counts.csv](https://github.com/neo4j/neo4j/files/13213334/counts.csv)

```
MATCH ()-[r]->()
RETURN type(r), count(*)
```
[relations.csv](https://github.com/neo4j/neo4j/files/13213373/relations.csv)

```
MATCH (a)-[r]->()
UNWIND labels(a) AS label
WITH label, type(r) as relType
RETURN [label, relType], count(*);
```
[label_rel.csv](https://github.com/neo4j/neo4j/files/13213406/label_rel.csv)


```
MATCH ()-[r]->(a)
UNWIND labels(a) AS label
WITH label, type(r) as relType
RETURN [relType, label], count(*);
```
[rel_label.csv](https://github.com/neo4j/neo4j/files/13213447/rel_label.csv)

**Comment by arnefischereit** [comment]
Hi @tpcgold,

Sorry, with `COUNT { () }` I meant `RETURN COUNT { () }` but the alternative you provided is equivalent.

I took note of your stats and will do some investigation on them.

Regards,

Arne

**Comment by arnefischereit** [comment]
Hi @tpcgold,

I have now registered your suggestion to pick the direction of expansion of VarExpand(Into) as a feature request. When discussing it, it was noted by the others that we do not have a lot of information in this case, so that we have a lot of uncertainty. I added your example to the feature request and we will consider it for future development.

Let me know if there is anything else I can do for you in this case. Otherwise, I will close this issue for now.

Regards,

Arne
